avalanche/training/templates/online_supervised.py

```diff
@@ -1,209 +1,332 @@
+        """
+        other_dataloader_args = {}
+
+        if parse_version(torch.__version__) >= parse_version("1.7.0"):
+            other_dataloader_args["persistent_workers"] = persistent_workers
+
+        self.dataloader = DataLoader(
+            self.adapted_dataset,
+            num_workers=num_workers,
+            batch_size=self.eval_mb_size,
+            pin_memory=pin_memory,
+            **other_dataloader_args
+        )
+
+    def forward(self):
+        """Compute the model's output given the current mini-batch."""
+        return avalanche_forward(self.model, self.mb_x, self.mb_task_id)
 
     def model_adaptation(self, model=None):
-        """Adapts the model to the data from the current
-           (full) experience.
+        """Adapts the model to the current data.
 
         Calls the :class:`~avalanche.models.DynamicModule`s adaptation.
         """
         if model is None:
             model = self.model
 
-        for module in model.modules():
-            if isinstance(module, DynamicModule):
-                module.adaptation(self.full_experience)
+        # For training:
+        if isinstance(self.experience, OnlineCLExperience):
+            # If the strategy has access to task boundaries, adapt the model
+            # for the whole origin experience to add the
+            if self.experience.access_task_boundaries:
+                avalanche_model_adaptation(model,
+                                           self.experience.origin_experience)
+            else:
+                self.model_params_before_adaptation = list(model.parameters())
+                avalanche_model_adaptation(model, self.experience)
+
+        # For evaluation, the experience is not necessarily an online
+        # experience:
+        else:
+            avalanche_model_adaptation(model, self.experience)
+
         return model.to(self.device)
+
+    def _unpack_minibatch(self):
+        """We assume mini-batches have the form <x, y, ..., t>.
+        This allows for arbitrary tensors between y and t.
+        Keep in mind that in the most general case mb_task_id is a tensor
+        which may contain different labels for each sample.
+        """
+        assert len(self.mbatch) >= 3
+        super()._unpack_minibatch()
+
+    def eval_dataset_adaptation(self, **kwargs):
+        """Initialize `self.adapted_dataset`."""
+        self.adapted_dataset = self.experience.dataset
+        self.adapted_dataset = self.adapted_dataset.eval()
+
+    def make_optimizer(self):
+        """Optimizer initialization.
+
+        Called before each training experience to configure the optimizer.
+        """
+        # We reset the optimizer's state after each experience if task
+        # boundaries are given, otherwise it updates the optimizer only if
+        # new parameters are added to the model after each adaptation step.
+
+        # We assume the current experience is an OnlineCLExperience:
+        if self.experience.access_task_boundaries:
```

Comment:
This condition is incomplete. It should be something like:
```
self.experience.access_task_boundaries and self.experience.is_first_subexperience
```

Can you add a test to make sure we get this right?

Reply:
The `access_task_boundaries` + `is_first_subepxerience` checking is done in `base_online_sgd` in line 156. There we first check both conditions to adapt the model and call `make_optimizer`. 

So here we only check task boundaries to check whether we need to reset (for task boundaries) or update (without task boundaries) the optimizer. Therefore, if the strategy has task boundaries, it means it was already checked in terms of being the first sub-exp inside online_base_sgd.

Reply:
ok, it makes sense. I expected the if inside the method but your solution is also correct.